### DISCUSSION TITLE
How to change .gltf morph targets on Android

### DISCUSSION THREAD

**Discussion by murielK** [google/filament#4094] [Q&A]
Hi everyone,

is it possible to change the morph targets on a .gltf file? 

I did lots of research about that and at first, I thought that it could be done via the FilamentAsset.animator property and I could just animate the morph targets by running or not the animator but for some reasons, the animationCount property is always returning 0.

After few more diggings and I found out that apparently, I need to use VertexBuffer and 1 or 2 of those VertexAttribute   
"      CUSTOM0,      // custom or MORPH_POSITION_0
        CUSTOM1,      // custom or MORPH_POSITION_1
        CUSTOM2,      // custom or MORPH_POSITION_2
        CUSTOM3,      // custom or MORPH_POSITION_3
        CUSTOM4,      // custom or MORPH_TANGENTS_0
        CUSTOM5,      // custom or MORPH_TANGENTS_1
        CUSTOM6,      // custom or MORPH_TANGENTS_2
        CUSTOM7       // custom or MORPH_TANGENTS_3"

 to change the morph targets but the issue is, I have no clue about how to use that VertexBuffer thing :/.

Can someone please show me an example on how to use VertexBuffer to change morph target using ModelViewer?

PS: This is my first time using Filament and 3D in general, so I apologize in advance for my newbie questions :D

**Comment by prideout** [comment]
So, you have a glTF file that already has morphing, and you already know how to play the animation that exists in the glTF file, correct?

You now wish to mutate the vertices in those morph targets at run time, right?  Currently our gltfio library does not expose the BufferObject objects that would allow you to do this.

If you're mutating the vertices very frequently, then why use morphing at all?  The goal of morphing is efficiency; for each frame, the GPU receives new weights rather than new vertex positions.

Setting aside glTF and morphing, note that Filament generally lets you mutate the contents of any VertexBuffer or BufferObject.  We have an example of this in ImGuiHelper.

**Reply by murielK** [reply]
We have glTF files that already have morphing, that is correct But I do not know how to play the animation.

Yes, that is correct, I wish to mutate the vertices in those morph targets at run time. Since your gltfio library does not expose the BufferObject object for that, is that mean it is not possible to update the morph targets at run time?


But maybe I said it wrong in the beginning, the point is not to mutate the vertices frequently, but only once when some .glTF files parts are loaded.

**Reply by prideout** [reply]
Let's tackle one thing at a time, since you don't know how to play the existing animation that's in the glTF file, let's start there.

Here's a snippet from the AssetLoader docs:

```
 * // Obtain the simple animation interface.
 * Animator* animator = asset->getAnimator();
 *
 * // Free the glTF hierarchy as it is no longer needed.
 * asset->releaseSourceData();
 *
 * // Add renderables to the scene.
 * scene->addEntities(asset->getEntities(), asset->getEntityCount());
 *
 * // Execute the render loop and play the first animation.
 * do {
 *      animator->applyAnimation(0, time);
 *      animator->updateBoneMatrices();
 *      if (renderer->beginFrame(swapChain)) {
 *          renderer->render(view);
 *          renderer->endFrame();
 *      }
 * } while (!quit);

```

Are you calling `applyAnimation` on every frame?

**Reply by murielK** [reply]
First of all thank you for your time and effort.

I'm using a custom ModelViewer, I was inspired by the one that is provided [here](https://github.com/google/filament/tree/main/android/samples/sample-gltf-viewer) because I needed to load multiple .glTF files.

About the animation, I already did something similar inside the doFrame(frameTimeNanos : Long) from Choreographer.FrameCallback (also inspired by the sample I mentioned above). But it is not working for me.

`                animator.apply {

                    val animationCount: Int = animationCount

                    logger.d("Animation count: $animationCount")

                    if (animationCount > 0){
                        val elapsedTimeSeconds =
                            (frameTimeNanos - startTime).toDouble() / 1_000_000_000
                        applyAnimation(0, elapsedTimeSeconds.toFloat())
                        updateBoneMatrices()
                    }
                }

                modelViewer.render(frameTimeNanos)`

Above is the animation code part inside the doFrame callback.

Apparently, from what the graphic designer said, I need to manually mutate the morph targets at run time for some .gltf files and it should be done only once.

So how can I do that? Is it possible do that with VertexBuffer if yes how, pretty please :)

**Reply by murielK** [reply]
@prideout  Can you please guide me with some code? On how can I mutate the morph targets at run time using VertexBuffer? if that's even possible.

**Reply by prideout** [reply]
Unless you're generating vertex positions programmatically, there's no need to mutate the VertexBuffer data manually.  If you're simply playing an animation that is defined in the glTF file, then `applyAnimation` should work.  Note that a single glTF asset might define multiple animations; since you're passing `0` you're selecting the first one.

Are you able to play the animation in your asset with another viewer like one these:

- https://sandbox.babylonjs.com/
- https://gltf-viewer.donmccurdy.com/

**Reply by murielK** [reply]
I try them in both links but the animation is not playing. The files will not play any animation by themself.

Sorry for the big confusion, but my .gltf file just contains morph targets. And I just need to modify the target values (from 0...to 1f) at run time.

**Reply by murielK** [reply]
@prideout Just for this context, my iOS mate can achieve that by changing the following property `Geometry.morpher.weight = 0...1`

**Reply by prideout** [reply]
Ah ok, in that case you can skip calling `applyAnimation`, and just use `RenderableManager::setMorphWeights()`.  You'll need to do this manually for every node in the asset that has a renderable component (see `FilamentAsset::getEntities`).

Note that it would be easier if the model itself defined an animation that varied the weights from 0 to 1 as time progressed from 0 to 1.  This would allow you to use applyAnimation rather than setting the weights on each individual node.

**Reply by murielK** [reply]
I use this right before i draw the frame 

  
```
  asset.entities.forEach { entity ->
            val rmInstance = renderableManager.getInstance(entity)
            renderableManager.setMorphWeights(
                rmInstance,
                if (asset.getName(entity).startsWith("target")) floatArrayOf(
                    1.0f,
                    0.0f,
                    0.0f,
                    0.0f
                ) else floatArrayOf(0.0f, 0.0f, 0.0f, 0.0f)
            )
        }
```


But it does not work, unfortunately. Please note that I used the renderable manager provided by the Engine itself.
```

    /**
     * Updates the vertex morphing weights on a renderable, all zeroes by default.
     *
     * <p>This is specified using a 4-tuple, one float per morph target. If the renderable has fewer
     * than 4 morph targets, then clients should fill the unused components with zeroes.</p>
     *
     * <p>The renderable must be built with morphing enabled.</p>
     *
     * @see Builder#morphing
     */
    public void setMorphWeights(@EntityInstance int i, @NonNull @Size(min = 4) float[] weights) {
        nSetMorphWeights(mNativeObject, i, weights);
    }
```

But as you can see above, the doc says that _The renderable must be built with morphing enabled_ .  I do not think that for this context, I have any control over this. 

What I'm missing here? :/

Again, thanks a lot for your amazing support so far.  I have learned so much in the few days since I started this discussion.

**Reply by prideout** [reply]
If the glTF provides vertex data for TARGET0, TARGET1, etc, then AssetLoader in gltfio will build the renderable with morphing enabled.

You can look for `builder.morphing(true)` in `AssetLoader.cpp` and double check that this codeline is being executed.  Does your model have more than 4 targets?  If so, it would be difficult to use the low-level `setMorphWeights` method.

How is your glTF generated?  Is it exported from Blender?  If you are working with a technical artist, I would ask the artist to include a simple animation in the glTF file that exercises the morphing functionality.

**Reply by murielK** [reply]
Our model have 2 targets only. 

glTF are exported from C4d and I have forwarded your recommendation to our technical artist. 

am I doing anything wrong in the code i have shared with you above? 

Again, thanks a lot for your support. ❤️

**Reply by prideout** [reply]
Interesting -- I assume the C4d exporter supports animation so the artist should be able to provide this.

The code you pasted looks fine.  For the final product, instead of `(1.0f, 0.0f, 0.0f, 0.0f)` I would expect your float array to be more like `(t, 1-t, 0.0f, 0.0f)`, depending where you want to be in the animation.

**Reply by nmfisher** [reply]
@prideout is there any sample code for using `setMorphWeights`? I've been struggling with this all day and I still haven't managed to get it working. I'm using the `third_party/models/AnimatedMorphCube/AnimatedMorphCube.glb` which definitely has two morph targets, I can trigger the animation fine (and I can manually move the morph targets outside Filament, e.g. https://gltf-viewer.donmccurdy.com/)
I've confirmed that `numMorphTargets` is non-zero and that `builder.morphing` is called, but the cube is always displayed at its original size.

I'm sure it's a problem on my end, I just can't see what I'm missing.

**Reply by prideout** [reply]
@nmfisher You can look at our MorphHelper class if you need an example of using setMorphWeights.  It's part of our gltfio library, which can handle AnimatedMorphCube just fine.

**Reply by nmfisher** [reply]
Thanks @prideout.  I revisited this today and as expected it was just a brainfart on my end - I was invoking `getFirstEntityByName` using the mesh name (i.e. "Cube"), when it should have been the node name ("AnimatedMorphCube").

**Reply by atul161** [reply]
@prideout  setting morph targest frequently slow the other ui component like horizontal recycler view.

Is there any workaround to have smooth experience of other ui elements.

I am doing all my calculations on io thread and updating ui on main thread. 

Still i am not able to achieve smooth transition on other ui elements.

**Reply by atul161** [reply]
@prideout  morph target is working but dark spot appears whenever  i am setting morph weights from framecallback.
This is worst behaviour and happening in almost every phone.  Please have a look on this! 

For now we have excluded this feature from our product.
